### DISCUSSION TITLE
Contributor sync notes - 2024-01-31

### DISCUSSION THREAD

**Discussion by tdikland** [delta-incubator/delta-sharing-rs#38] [General]
These are rough notes of the biweekly contributor sync. Please contact @MrPowers if you want to be included in the invitation.

@linzhou-db:
- This working group is focused on building a delta sharing server implementation. However it is interesting to understand how clients/connectors are build. Especially the rust -> python bindings (e.g. with PyO3)
- Servers, clients and connectors should leverage the delta-kernel-rs project

@roeap:
- The delta-rs (the Delta Lake rust ecosystem) is preparing for the public release of delta-kernel-rs
- delta-kernel-rs has the protocol implementation and interfaces for the so-called engine to do the work (IO, expression evaluation, etc.)
- It's based on a Arrow memory model
- Whenever operations are done on a delta table, one typically needs a query engine to do so. Datafusion is often the engine of choice
- A Delta Table shared with Delta Sharing is basically a read-only delta table. It may be beneficial to model it as such into the delta-rs ecosystem.
- The main task of the Delta Sharing server is to replay the delta log.

Thoughts around the roadmap for the coming months:
- Build a library crate that implements the Delta Sharing protocol
- Allow for ways to extend the core protocol
	- One can define extension points in the protocol implementation (e.g. a `Catalog` abstraction to fetch share/schema/table information). The crate should define default implementations, but be extensible for customisations.
	- One can define extra routes that are not part of protocol. Think about an admin route for managing shares/schemas/tables or a discovery route for displaying rich metadata about the shared assets.
	- One should be able to easily add RateLimiting, Logging, etc. to the server.
- The code needs a bit of cleanup/refactoring to reach the goals above. Currently the project has coupled some of the concepts above, which makes it hard to introduce the extension points envisioned.

Actions:
- (Robert) Share a design document to asynchronously collaborate on the question of how to organize the project

**Comment by dennyglee** [comment]
Thanks for getting the notes down @tdikland.  @roeap whenever you're ready, let me know so I can help with the design doc as well.  I'm sure most of us will want to help, too.  :)

**Comment by MrPowers** [comment]
@tdikland - thanks so much for these detailed notes!!!
